### ISSUE TITLE
DxeImageVerificationHandler() fails open in case of dbx signature check (Bugzilla Bug 1608)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#7825]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 1608
Date: 2019-03-10T21:08:46+00:00
From: Wang, Jian J &lt;<jian.j.wang>&gt;
To:   Wang, Jian J &lt;<jian.j.wang>&gt;
CC:   @corthon, @dickwilkins, eugene, f4bug, @jkmathews, kimolsun2020, @lersek, @mdkinney, rschiron, @vincent-j-zimmer

Last updated: 2021-09-22T15:47:02+00:00

**Comment by tianocore-issues** [comment]
## Comment 8039
Date: 2019-06-13 17:05:39 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Created attachment 357
v2

Attachment: [BZ-357-0001-SecurityPkg-DxeImageVerificationLib-fix-false-negati.patch](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-357-0001-SecurityPkg-DxeImageVerificationLib-fix-false-negati.patch)

**Comment by tianocore-issues** [comment]
## Comment 8186
Date: 2019-06-25 16:49:15 +0000
From: @corthon

Patch v2 looks good. What’s the timeline for release?

**Comment by tianocore-issues** [comment]
## Comment 8307
Date: 2019-07-01 20:02:21 +0000
From: @dickwilkins

Phoenix would also like to know if this patch is ready to be rolled out to our OEM customers. Also, has the embargo time clock started running? (Is this the &quot;official&quot; patch?)

**Comment by tianocore-issues** [comment]
## Comment 8457
Date: 2019-07-15 12:30:57 +0000
From: @corthon

Upon integration, there&apos;s a couple of problems with this patch.

1) In IsForbiddenByDbx(), IsForbidden needs to be set to FALSE after the final for-loop, otherwise the TRUE default will reject anything that wasn&apos;t found as a hash.
2) In IsAllowedByDb(), VerifyStatus needs to be set to TRUE after IsCertHashFoundInDatabase() in the case where there was no error, but IsFound == FALSE.

These changes can be seen in the following commit in our Mu security patch repo. Currently, anyone with permission to view the TianoCore security bugs is welcome to request access to this repo. We intend to maintain it going forward as each Mu release stabilizes.

https://dev.azure.com/projectmu/Mu_Security_Private_Release/_git/mu_basecore/commit/d5b1dc4e8c1da98c2adea64ccd0208e75535b0db

If you can&apos;t view the commit, email me (bret.barkelew@microsoft.com) to request access.

**Comment by tianocore-issues** [comment]
## Comment 8538
Date: 2019-07-22 11:13:32 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Bret and Dick,

Per current process, once an issue has solution (patch tested and attached), there&apos;ll be a 6-month embargo period before disclosing. I&apos;d assume the embargo starting from June.

Jian

**Comment by tianocore-issues** [comment]
## Comment 8539
Date: 2019-07-22 11:14:59 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Bret Barkelew from comment #6)
&gt; Upon integration, there&apos;s a couple of problems with this patch.
&gt; 
&gt; 1) In IsForbiddenByDbx(), IsForbidden needs to be set to FALSE after the
&gt; final for-loop, otherwise the TRUE default will reject anything that wasn&apos;t
&gt; found as a hash.
&gt; 2) In IsAllowedByDb(), VerifyStatus needs to be set to TRUE after
&gt; IsCertHashFoundInDatabase() in the case where there was no error, but
&gt; IsFound == FALSE.
&gt; 
&gt; These changes can be seen in the following commit in our Mu security patch
&gt; repo. Currently, anyone with permission to view the TianoCore security bugs
&gt; is welcome to request access to this repo. We intend to maintain it going
&gt; forward as each Mu release stabilizes.
&gt; 
&gt; https://dev.azure.com/projectmu/Mu_Security_Private_Release/_git/mu_basecore/
&gt; commit/d5b1dc4e8c1da98c2adea64ccd0208e75535b0db
&gt; 
&gt; If you can&apos;t view the commit, email me (bret.barkelew@microsoft.com) to
&gt; request access.

Thanks for catching the problems. I&apos;ve sent you email for the repo access. Once I confirmed them I&apos;ll upload the v3 patch here.

**Comment by tianocore-issues** [comment]
## Comment 8565
Date: 2019-07-24 21:34:52 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Bret Barkelew from comment #6)
&gt; Upon integration, there&apos;s a couple of problems with this patch.
&gt; 
&gt; 1) In IsForbiddenByDbx(), IsForbidden needs to be set to FALSE after the
&gt; final for-loop, otherwise the TRUE default will reject anything that wasn&apos;t
&gt; found as a hash.

I think IsForbidden should not be set to FALSE after final for-loop. If the function can run to the final for-loop, the IsForbidden must be FALSE. If the AuthData data can pass timestamp check, the IsForbidden is set to FALSE and code logic can reach the point after for-loop. Otherwise, the code logic just jump to Done label. So there&apos;s no need to set it to FALSE after the for-loop, unless you think the CertNumber==0 is a valid situation.

**Comment by tianocore-issues** [comment]
## Comment 8566
Date: 2019-07-25 02:16:42 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Created attachment 368
v3

v3 changes:
  a. incorporate Bret&apos;s fixes
  b. fix two more issues:
     1) check CertNumber against zero in IsForbiddenByDbx()
     2) init DbxDataSize to 0 before testing the variable size in IsAllowedByDb()

Attachment: [BZ-368-0001-SecurityPkg-DxeImageVerificationLib-fix-false-negati.patch](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-368-0001-SecurityPkg-DxeImageVerificationLib-fix-false-negati.patch)

**Comment by tianocore-issues** [comment]
## Comment 8567
Date: 2019-07-25 02:19:03 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Wang, Jian J from comment #9)
&gt; (In reply to Bret Barkelew from comment #6)
&gt; &gt; Upon integration, there&apos;s a couple of problems with this patch.
&gt; &gt; 
&gt; &gt; 1) In IsForbiddenByDbx(), IsForbidden needs to be set to FALSE after the
&gt; &gt; final for-loop, otherwise the TRUE default will reject anything that wasn&apos;t
&gt; &gt; found as a hash.
&gt; 
&gt; I think IsForbidden should not be set to FALSE after final for-loop. If the
&gt; function can run to the final for-loop, the IsForbidden must be FALSE. If
&gt; the AuthData data can pass timestamp check, the IsForbidden is set to FALSE
&gt; and code logic can reach the point after for-loop. Otherwise, the code logic
&gt; just jump to Done label. So there&apos;s no need to set it to FALSE after the
&gt; for-loop, unless you think the CertNumber==0 is a valid situation.

Forget my above comment. Your change is correct. I uploaded v3 patch, which includes two more issues we found besides your changes.

**Comment by tianocore-issues** [comment]
## Comment 8572
Date: 2019-07-25 17:00:03 +0000
From: @corthon

Thanks!

**Comment by tianocore-issues** [comment]
## Comment 8699
Date: 2019-08-01 15:57:48 +0000
From: @lersek

Jian,

can you please assign a CVE number to this vulnerability?

And once you have the CVE ID, can you please add it to the Alias field on this bugzilla ticket?

Thank you.

**Comment by tianocore-issues** [comment]
## Comment 8700
Date: 2019-08-01 15:59:41 +0000
From: @lersek

This issue affects all OvmfPkg platforms that are built with -D SECURE_BOOT_ENABLE.

The issue doesn&apos;t affect ArmVirtPkg platforms (as none of those support secure boot yet).

**Comment by tianocore-issues** [comment]
## Comment 8701
Date: 2019-08-01 16:04:02 +0000
From: @dickwilkins

Is this patch now considered &quot;the solution&quot;? And the embargo period starts from when v3 was available? That is Jul 25, 2019 thru Jan 25, 2020.

**Comment by tianocore-issues** [comment]
## Comment 8743
Date: 2019-08-02 06:26:54 +0000
From: @lersek

*** Bug 1945 has been marked as a duplicate of this bug. ***

**Comment by tianocore-issues** [comment]
## Comment 8776
Date: 2019-08-02 18:40:21 +0000
From: @lersek

(actually I&apos;m going to remove the RHBZ downstream tracker from See Also -- not necessarily because I want to keep that information hidden (the RHBZ is private anyway), but because vendor BZ references just clutter the See Also field, they are useless to all TianoCore BZ users except the one vendor in question)

**Comment by tianocore-issues** [comment]
## Comment 9443
Date: 2019-09-03 07:53:43 +0000
From: @lersek

(In reply to Wang, Jian J from comment #10)
&gt; Created attachment 368 [details]
&gt; v3

My comments:

&gt;| diff --git a/SecurityPkg/Library/DxeImageVerificationLib/DxeImageVerificationLib.c b/SecurityPkg/Library/DxeImageVerificationLib/DxeImageVerificationLib.c
&gt;| index fe4cdccedd..de1b39166a 100644
&gt;| --- a/SecurityPkg/Library/DxeImageVerificationLib/DxeImageVerificationLib.c
&gt;| +++ b/SecurityPkg/Library/DxeImageVerificationLib/DxeImageVerificationLib.c
&gt;| @@ -821,22 +821,23 @@ AddImageExeInfo (
&gt;|    @param[in]  SignatureList     Pointer to the Signature List in forbidden database.
&gt;|    @param[in]  SignatureListSize Size of Signature List.
&gt;|    @param[out] RevocationTime    Return the time that the certificate was revoked.
&gt;| +  @param[in]  IsFound           Search result.

(1) &quot;IsFound&quot; is an output parameter, not an input parameter.

(2) Please document explicitly that the caller should only consult
&quot;IsFound&quot; in case EFI_SUCCESS is returned.

&gt;|
&gt;| -  @return TRUE   The certificate hash is found in the forbidden database.
&gt;| -  @return FALSE  The certificate hash is not found in the forbidden database.
&gt;| +  @return EFI_SUCCESS           Finished the search without any error.
&gt;| +  @return Others                Error occurred in the search of database.

(3) The EFI_SUCCESS return value should not be documented with @return;
it should be @retval.

&gt;|    //
&gt;|    // Retrieve the TBSCertificate from the X.509 Certificate.
&gt;|    //
&gt;|    if (!X509GetTBSCert (Certificate, CertSize, &amp;TBSCert, &amp;TBSCertSize)) {
&gt;| -    return FALSE;
&gt;| +    goto Done;
&gt;|    }

(4) I think the &quot;goto Done&quot; is unjustified here. It needlessly
complicates the control flow. At this point we have *certainly* not
allocated &quot;HashCtx&quot;. Therefore we should simply write &quot;return Status&quot;.
(Status is set to EFI_ABORTED at this point.)

&gt;| @@ -894,16 +896,13 @@ IsCertHashFoundInDatabase (
&gt;|      if (HashCtx == NULL) {
&gt;|        goto Done;
&gt;|      }
&gt;| -    Status = mHash[HashAlg].HashInit (HashCtx);
&gt;| -    if (!Status) {
&gt;| +    if (!mHash[HashAlg].HashInit (HashCtx)) {
&gt;|        goto Done;
&gt;|      }
&gt;| -    Status = mHash[HashAlg].HashUpdate (HashCtx, TBSCert, TBSCertSize);
&gt;| -    if (!Status) {
&gt;| +    if (!mHash[HashAlg].HashUpdate (HashCtx, TBSCert, TBSCertSize)) {
&gt;|        goto Done;
&gt;|      }
&gt;| -    Status = mHash[HashAlg].HashFinal (HashCtx, CertDigest);
&gt;| -    if (!Status) {
&gt;| +    if (!mHash[HashAlg].HashFinal (HashCtx, CertDigest)) {
&gt;|        goto Done;
&gt;|      }
&gt;|

(5) Here we have an independent (preexistent) bug. We should fix it --
in a separate patch.

Namely, we call AllocatePool() in the loop, but we don&apos;t free &quot;HashCtx&quot;
in the loop!

Therefore, right after the HashFinal() call succeeds, we should do:

  FreePool (HashCtx);
  HashCtx = NULL;

&gt;| @@ -934,12 +934,14 @@ IsCertHashFoundInDatabase (
&gt;|      DbxList  = (EFI_SIGNATURE_LIST *) ((UINT8 *) DbxList + DbxList-&gt;SignatureListSize);
&gt;|    }
&gt;|
&gt;| +  Status   = EFI_SUCCESS;
&gt;| +

(6) The extra whitespace is unwarranted for, in this case.

&gt;|  Done:
&gt;|    if (HashCtx != NULL) {
&gt;|      FreePool (HashCtx);
&gt;|    }
&gt;|
&gt;| -  return IsFound;
&gt;| +  return Status;
&gt;|  }
&gt;|
&gt;|  /**

Other than those cosmetic problems, the updates to
IsCertHashFoundInDatabase() look OK to me.

&gt;| @@ -949,17 +951,19 @@ Done:
&gt;|    @param[in]  Signature           Pointer to signature that is searched for.
&gt;|    @param[in]  CertType            Pointer to hash algrithom.
&gt;|    @param[in]  SignatureSize       Size of Signature.
&gt;| +  @param[in]  IsFound             Search result.

(7) Same as (1).

(8) Same as (2).

&gt;|
&gt;| -  @return TRUE                    Found the signature in the variable database.
&gt;| -  @return FALSE                   Not found the signature in the variable database.
&gt;| +  @return EFI_SUCCESS             Finished the search without any error.
&gt;| +  @return Others                  Error occurred in the search of database.

(9) Same as (3).


&gt;|  **/
&gt;| -BOOLEAN
&gt;| +EFI_STATUS
&gt;|  IsSignatureFoundInDatabase (
&gt;| -  IN CHAR16             *VariableName,
&gt;| -  IN UINT8              *Signature,
&gt;| -  IN EFI_GUID           *CertType,
&gt;| -  IN UINTN              SignatureSize
&gt;| +  IN  CHAR16            *VariableName,
&gt;| +  IN  UINT8             *Signature,
&gt;| +  IN  EFI_GUID          *CertType,
&gt;| +  IN  UINTN             SignatureSize,
&gt;| +  OUT BOOLEAN           *IsFound
&gt;|    )

(10) The IsSignatureFoundInDatabase() update (and the call site updates)
should be split to a separate patch, from the
IsCertHashFoundInDatabase() patch.

Larger patches are more difficult to review, and both functions can be
isolated from each other.

&gt;|  {
&gt;|    EFI_STATUS          Status;
&gt;| @@ -969,22 +973,29 @@ IsSignatureFoundInDatabase (
&gt;|    UINT8               *Data;
&gt;|    UINTN               Index;
&gt;|    UINTN               CertCount;
&gt;| -  BOOLEAN             IsFound;
&gt;|
&gt;|    //
&gt;|    // Read signature database variable.
&gt;|    //
&gt;| -  IsFound   = FALSE;
&gt;| +  *IsFound  = FALSE;
&gt;|    Data      = NULL;
&gt;|    DataSize  = 0;
&gt;|    Status    = gRT-&gt;GetVariable (VariableName, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DataSize, NULL);
&gt;|    if (Status != EFI_BUFFER_TOO_SMALL) {
&gt;| -    return FALSE;
&gt;| +    if (Status == EFI_NOT_FOUND) {
&gt;| +      //
&gt;| +      // No database, no need to search, or search has been done already.

(11) I don&apos;t understand the &quot;search has been done already&quot; part -- why
does the EFI_NOT_FOUND return code imply an earlier search?

&gt;| +      //
&gt;| +      Status = EFI_SUCCESS;
&gt;| +    } else {
&gt;| +      Status = EFI_ABORTED;
&gt;| +    }
&gt;| +    goto Done;
&gt;|    }

(12) How about the following instead:

  Status = gRT-&gt;GetVariable (...);
  ASSERT (EFI_ERROR (Status));
  if (Status != EFI_BUFFER_TOO_SMALL) {
    if (Status == EFI_NOT_FOUND) {
      //
      // No database, no need to search
      //
      Status = EFI_SUCCESS;
    }
    return Status;
  }

Due to three considerations:

- with DataSize=0 on input, gRT-&gt;GetVariable() can never succeed,

- if gRT-&gt;GetVariable() fails with an error code that is different from
  both EFI_BUFFER_TOO_SMALL and EFI_NOT_FOUND, then we don&apos;t need to
  overwrite that error code with EFI_ABORTED; we can just propagate the
  error outwards. The interface contract forbids the caller from
  consuming &quot;IsFound&quot; on any error.

- We haven&apos;t allocated anything yet; no need for complicating the
  control flow with a &quot;goto&quot;.

If you disagree, I won&apos;t insist -- the v3 code above is not wrong, it&apos;s
just a bit too hard to reason about (at least for me).

&gt;|
&gt;|    Data = (UINT8 *) AllocateZeroPool (DataSize);
&gt;|    if (Data == NULL) {
&gt;| -    return FALSE;
&gt;| +    goto Done;
&gt;|    }

(13) Again, not a bug per se, just difficult to reason about -- with the
&quot;goto&quot; here, we rely on &quot;Status&quot; being set to EFI_BUFFER_TOO_SMALL, by
gRT-&gt;GetVariable().

In addition, the &quot;goto&quot; is unnecessary in the first place; we haven&apos;t
allocated anything yet.

I think it would be cleaner to write:

  return EFI_OUT_OF_RESOURCES;

&gt;|
&gt;|    Status = gRT-&gt;GetVariable (VariableName, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DataSize, Data);
&gt;| @@ -1004,22 +1015,18 @@ IsSignatureFoundInDatabase (
&gt;|            //
&gt;|            // Find the signature in database.
&gt;|            //
&gt;| -          IsFound = TRUE;
&gt;| +          *IsFound = TRUE;
&gt;|            //
&gt;|            // Entries in UEFI_IMAGE_SECURITY_DATABASE that are used to validate image should be measured
&gt;|            //
&gt;|            if (StrCmp(VariableName, EFI_IMAGE_SECURITY_DATABASE) == 0) {
&gt;|              SecureBootHook (VariableName, &amp;gEfiImageSecurityDatabaseGuid, CertList-&gt;SignatureSize, Cert);
&gt;|            }
&gt;| -          break;
&gt;| +          goto Done;
&gt;|          }
&gt;|
&gt;|          Cert = (EFI_SIGNATURE_DATA *) ((UINT8 *) Cert + CertList-&gt;SignatureSize);
&gt;|        }
&gt;| -
&gt;| -      if (IsFound) {
&gt;| -        break;
&gt;| -      }
&gt;|      }
&gt;|

(14) While the code is not wrong, I disagee about replacing the &quot;break&quot;
statements with &quot;goto&quot;. For two reasons:

- This change is orthogonal to the bug being fixed, therefore it should
  be in a separate patch, minimally. (The same replacement could be done
  using the original &quot;IsFound&quot; BOOLEAN variable.)

- According to the edk2 C Coding Standards, &quot;goto&quot; should be used for
  error handling, at the most. (I realize that the other function,
  IsCertHashFoundInDatabase(), breaks that rule even before this patch.)

If you really value consistency with IsCertHashFoundInDatabase() over
the CCS, then at least split this change to a separate patch, please.

&gt;|      DataSize -= CertList-&gt;SignatureListSize;
&gt;| @@ -1031,7 +1038,7 @@ Done:
&gt;|      FreePool (Data);
&gt;|    }
&gt;|
&gt;| -  return IsFound;
&gt;| +  return Status;
&gt;|  }
&gt;|
&gt;|  /**
&gt;| @@ -1212,6 +1219,7 @@ IsForbiddenByDbx (
&gt;|  {
&gt;|    EFI_STATUS                Status;
&gt;|    BOOLEAN                   IsForbidden;
&gt;| +  BOOLEAN                   IsFound;
&gt;|    UINT8                     *Data;
&gt;|    UINTN                     DataSize;
&gt;|    EFI_SIGNATURE_LIST        *CertList;
&gt;| @@ -1233,7 +1241,7 @@ IsForbiddenByDbx (
&gt;|    //
&gt;|    // Variable Initialization
&gt;|    //
&gt;| -  IsForbidden       = FALSE;
&gt;| +  IsForbidden       = TRUE;
&gt;|    Data              = NULL;
&gt;|    CertList          = NULL;
&gt;|    CertData          = NULL;
&gt;| @@ -1251,16 +1259,22 @@ IsForbiddenByDbx (
&gt;|    DataSize = 0;
&gt;|    Status   = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DataSize, NULL);
&gt;|    if (Status != EFI_BUFFER_TOO_SMALL) {
&gt;| -    return IsForbidden;
&gt;| +    if (Status == EFI_NOT_FOUND) {
&gt;| +      //
&gt;| +      // Apparently not in dbx if the database doesn&apos;t exist.
&gt;| +      //
&gt;| +      IsForbidden = FALSE;
&gt;| +    }
&gt;| +    goto Done;
&gt;|    }

(15) How about the following instead:

  Status = gRT-&gt;GetVariable (...);
  ASSERT (EFI_ERROR (Status));
  if (Status != EFI_BUFFER_TOO_SMALL) {
    if (Status == EFI_NOT_FOUND) {
      //
      // Evidently not in dbx if the database doesn&apos;t exist.
      //
      IsForbidden = FALSE;
    }
    return IsForbidden;
  }

Due to the following considerations:

- With DataSize=0, GetVariable cannot succeed. Be clear about that.

- We haven&apos;t allocated anything yet, so no need to jump to the Done
  label. The goto needlessly complicates the control flow.

- According to the English wiktionary, in modern speech, &quot;apparently&quot;
  primarily means &quot;seemingly&quot;. What you mean is &quot;evidently&quot; (which is
  only an archaic meaning of the word &quot;apparently&quot;).
  &lt;https://en.wiktionary.org/wiki/apparently&gt;

Again, the code is not wrong, but it could be improved for clarity&apos;s
sake.

&gt;|    Data = (UINT8 *) AllocateZeroPool (DataSize);
&gt;|    if (Data == NULL) {
&gt;| -    return IsForbidden;
&gt;| +    goto Done;
&gt;|    }

(16) This change is unneeded, en bloc. We haven&apos;t allocated anything
yet.

&gt;|
&gt;|    Status = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DataSize, (VOID *) Data);
&gt;|    if (EFI_ERROR (Status)) {
&gt;| -    return IsForbidden;
&gt;| +    goto Done;
&gt;|    }
&gt;|
&gt;|    //
&gt;| @@ -1322,7 +1336,7 @@ IsForbiddenByDbx (
&gt;|    //       UINT8  Certn[];
&gt;|    //
&gt;|    Pkcs7GetSigners (AuthData, AuthDataSize, &amp;CertBuffer, &amp;BufferLength, &amp;TrustedCert, &amp;TrustedCertLength);
&gt;| -  if ((BufferLength == 0) || (CertBuffer == NULL)) {
&gt;| +  if ((BufferLength == 0) || (CertBuffer == NULL) || (*CertBuffer) == 0) {
&gt;|      IsForbidden = TRUE;
&gt;|      goto Done;
&gt;|    }

(17) The nullity check on (*CertBuffer) seems entirely orthogonal to the
present patch. I can&apos;t tell off-hand if this change is valid and/or
necessary, but minimally, it should be split to a separate patch.


&gt;| @@ -1340,12 +1354,13 @@ IsForbiddenByDbx (
&gt;|      //
&gt;|      CertPtr = CertPtr + sizeof (UINT32) + CertSize;
&gt;|
&gt;| -    if (IsCertHashFoundInDatabase (Cert, CertSize, (EFI_SIGNATURE_LIST *)Data, DataSize, &amp;RevocationTime)) {
&gt;| +    Status = IsCertHashFoundInDatabase (Cert, CertSize, (EFI_SIGNATURE_LIST *)Data, DataSize, &amp;RevocationTime, &amp;IsFound);
&gt;| +    if (EFI_ERROR (Status) || IsFound) {
&gt;|        //
&gt;|        // Check the timestamp signature and signing time to determine if the image can be trusted.
&gt;|        //
&gt;|        IsForbidden = TRUE;
&gt;| -      if (PassTimestampCheck (AuthData, AuthDataSize, &amp;RevocationTime)) {
&gt;| +      if (IsFound &amp;&amp; PassTimestampCheck (AuthData, AuthDataSize, &amp;RevocationTime)) {
&gt;|          IsForbidden = FALSE;
&gt;|          //
&gt;|          // Pass DBT check. Continue to check other certs in image signer&apos;s cert list against DBX, DBT

(18) The change is incorrect. We may have reached that location due to:

(a) IsCertHashFoundInDatabase() failing, or

(b) IsCertHashFoundInDatabase() succeeding, and setting &quot;IsFound&quot; to
true.

In case (a), we must not consume &quot;IsFound&quot;. Therefore, the condition
should be written as:

  (!EFI_ERROR (Status) &amp;&amp;
   PassTimestampCheck (AuthData, AuthDataSize, &amp;RevocationTime))


&gt;| @@ -1355,9 +1370,10 @@ IsForbiddenByDbx (
&gt;|        DEBUG ((DEBUG_INFO, &quot;DxeImageVerificationLib: Image is signed but signature failed the timestamp check.\n&quot;));
&gt;|        goto Done;
&gt;|      }
&gt;| -
&gt;|    }
&gt;|

(19) The above whitespace change is spurious, please undo it.

&gt;| +  IsForbidden = FALSE;
&gt;| +
&gt;|  Done:
&gt;|    if (Data != NULL) {
&gt;|      FreePool (Data);
&gt;| @@ -1388,6 +1404,7 @@ IsAllowedByDb (
&gt;|  {
&gt;|    EFI_STATUS                Status;
&gt;|    BOOLEAN                   VerifyStatus;
&gt;| +  BOOLEAN                   IsFound;
&gt;|    EFI_SIGNATURE_LIST        *CertList;
&gt;|    EFI_SIGNATURE_DATA        *CertData;
&gt;|    UINTN                     DataSize;
&gt;| @@ -1449,11 +1466,20 @@ IsAllowedByDb (
&gt;|                             mImageDigestSize
&gt;|                             );
&gt;|            if (VerifyStatus) {
&gt;| +            VerifyStatus = FALSE;
&gt;| +
&gt;|              //
&gt;|              // Here We still need to check if this RootCert&apos;s Hash is revoked
&gt;|              //
&gt;| +            DbxDataSize = 0;
&gt;|              Status   = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DbxDataSize, NULL);
&gt;| -            if (Status == EFI_BUFFER_TOO_SMALL) {
&gt;| +            if (Status != EFI_BUFFER_TOO_SMALL) {
&gt;| +              if (Status == EFI_NOT_FOUND) {
&gt;| +                //
&gt;| +                // Apparently not in dbx if the database doesn&apos;t exist.
&gt;| +                //
&gt;| +                VerifyStatus = TRUE;
&gt;| +              }
&gt;|                goto Done;
&gt;|              }

(20) I&apos;d again propose the pattern from (12) and (15).

&gt;|              DbxData = (UINT8 *) AllocateZeroPool (DbxDataSize);
&gt;| @@ -1466,7 +1492,8 @@ IsAllowedByDb (
&gt;|                goto Done;
&gt;|              }
&gt;|
&gt;| -            if (IsCertHashFoundInDatabase (RootCert, RootCertSize, (EFI_SIGNATURE_LIST *)DbxData, DbxDataSize, &amp;RevocationTime)) {
&gt;| +            Status = IsCertHashFoundInDatabase (RootCert, RootCertSize, (EFI_SIGNATURE_LIST *)DbxData, DbxDataSize, &amp;RevocationTime, &amp;IsFound);
&gt;| +            if (!EFI_ERROR (Status) &amp;&amp; IsFound) {
&gt;|                //
&gt;|                // Check the timestamp signature and signing time to determine if the RootCert can be trusted.
&gt;|                //
&gt;| @@ -1474,6 +1501,8 @@ IsAllowedByDb (
&gt;|                if (!VerifyStatus) {
&gt;|                  DEBUG ((DEBUG_INFO, &quot;DxeImageVerificationLib: Image is signed and signature is accepted by DB, but its root cert failed the timestamp check.\n&quot;));
&gt;|                }
&gt;| +            } else if (!EFI_ERROR (Status) &amp;&amp; !IsFound) {
&gt;| +              VerifyStatus = TRUE;
&gt;|              }
&gt;|
&gt;|              goto Done;

(21) The branching is too complex. After IsCertHashFoundInDatabase()
returns, you can check &quot;Status&quot; in itself. For example:

  Status = IsCertHashFoundInDatabase
             RootCert,
             RootCertSize,
             (EFI_SIGNATURE_LIST *)DbxData,
             DbxDataSize,
             &amp;RevocationTime,
             &amp;IsFound
             );
  if (EFI_ERROR (Status)) {
    goto Done;
  }

  if (!IsFound) {
    VerifyStatus = TRUE;
    goto Done;
  }

  VerifyStatus = PassTimestampCheck (
                   AuthData,
                   AuthDataSize,
                   &amp;RevocationTime
                   );
  if (!VerifyStatus) {
    DEBUG ((...));
  }
  goto Done;

&gt;| @@ -1580,6 +1609,7 @@ DxeImageVerificationHandler (
&gt;|    EFI_IMAGE_DATA_DIRECTORY             *SecDataDir;
&gt;|    UINT32                               OffSet;
&gt;|    CHAR16                               *NameStr;
&gt;| +  BOOLEAN                              IsFound;
&gt;|
&gt;|    SignatureList     = NULL;
&gt;|    SignatureListSize = 0;
&gt;| @@ -1732,15 +1762,32 @@ DxeImageVerificationHandler (
&gt;|        goto Done;
&gt;|      }
&gt;|
&gt;| -    if (IsSignatureFoundInDatabase (EFI_IMAGE_SECURITY_DATABASE1, mImageDigest, &amp;mCertType, mImageDigestSize)) {
&gt;| +    IsFound = TRUE;

(22) This assignment is not necessary, and its presence is confusing.
Please remove it.

&gt;| +    Status = IsSignatureFoundInDatabase (
&gt;| +               EFI_IMAGE_SECURITY_DATABASE1,
&gt;| +               mImageDigest,
&gt;| +               &amp;mCertType,
&gt;| +               mImageDigestSize,
&gt;| +               &amp;IsFound
&gt;| +               );
&gt;| +    if (EFI_ERROR (Status) || IsFound) {
&gt;|        //
&gt;|        // Image Hash is in forbidden database (DBX).
&gt;|        //
&gt;|        DEBUG ((DEBUG_INFO, &quot;DxeImageVerificationLib: Image is not signed and %s hash of image is forbidden by DBX.\n&quot;, mHashTypeStr));
&gt;| +      Status = EFI_ACCESS_DENIED;
&gt;|        goto Done;
&gt;|      }
&gt;|
&gt;| -    if (IsSignatureFoundInDatabase (EFI_IMAGE_SECURITY_DATABASE, mImageDigest, &amp;mCertType, mImageDigestSize)) {
&gt;| +    IsFound = FALSE;

(23) Same as (22).

&gt;| +    Status = IsSignatureFoundInDatabase (
&gt;| +               EFI_IMAGE_SECURITY_DATABASE,
&gt;| +               mImageDigest,
&gt;| +               &amp;mCertType,
&gt;| +               mImageDigestSize,
&gt;| +               &amp;IsFound
&gt;| +               );
&gt;| +    if (!EFI_ERROR (Status) &amp;&amp; IsFound) {
&gt;|        //
&gt;|        // Image Hash is in allowed database (DB).
&gt;|        //
&gt;| @@ -1751,6 +1798,7 @@ DxeImageVerificationHandler (
&gt;|      // Image Hash is not found in both forbidden and allowed database.
&gt;|      //
&gt;|      DEBUG ((DEBUG_INFO, &quot;DxeImageVerificationLib: Image is not signed and %s hash of image is not found in DB/DBX.\n&quot;, mHashTypeStr));
&gt;| +    Status = EFI_ACCESS_DENIED;
&gt;|      goto Done;
&gt;|    }
&gt;|
&gt;| @@ -1828,16 +1876,33 @@ DxeImageVerificationHandler (
&gt;|      //
&gt;|      // Check the image&apos;s hash value.
&gt;|      //
&gt;| -    if (IsSignatureFoundInDatabase (EFI_IMAGE_SECURITY_DATABASE1, mImageDigest, &amp;mCertType, mImageDigestSize)) {
&gt;| +    IsFound = TRUE;

(24) Same as (22).

&gt;| +    Status = IsSignatureFoundInDatabase (
&gt;| +               EFI_IMAGE_SECURITY_DATABASE1,
&gt;| +               mImageDigest,
&gt;| +               &amp;mCertType,
&gt;| +               mImageDigestSize,
&gt;| +               &amp;IsFound
&gt;| +               );
&gt;| +    if (EFI_ERROR (Status) || IsFound) {
&gt;|        Action = EFI_IMAGE_EXECUTION_AUTH_SIG_FOUND;
&gt;|        DEBUG ((DEBUG_INFO, &quot;DxeImageVerificationLib: Image is signed but %s hash of image is found in DBX.\n&quot;, mHashTypeStr));
&gt;|        VerifyStatus = EFI_ACCESS_DENIED;
&gt;|        break;
&gt;|      } else if (EFI_ERROR (VerifyStatus)) {
&gt;| -      if (IsSignatureFoundInDatabase (EFI_IMAGE_SECURITY_DATABASE, mImageDigest, &amp;mCertType, mImageDigestSize)) {
&gt;| +      IsFound = FALSE;

(25) Same as (22).

&gt;| +      Status = IsSignatureFoundInDatabase (
&gt;| +                 EFI_IMAGE_SECURITY_DATABASE,
&gt;| +                 mImageDigest,
&gt;| +                 &amp;mCertType,
&gt;| +                 mImageDigestSize,
&gt;| +                 &amp;IsFound
&gt;| +                 );
&gt;| +      if (!EFI_ERROR (Status) &amp;&amp; IsFound) {
&gt;|          VerifyStatus = EFI_SUCCESS;
&gt;|        } else {
&gt;|          DEBUG ((DEBUG_INFO, &quot;DxeImageVerificationLib: Image is signed but signature is not allowed by DB and %s hash of image is not found in DB/DBX.\n&quot;, mHashTypeStr));
&gt;| +        Status = EFI_ACCESS_DENIED;

(26) Why is the last assignment needed?

Before the patch, the value of &quot;Status&quot;, near the DEBUG macro
invocation, is guaranteed to be EFI_SUCCESS. See the HashPeImageByType()
invocation higher up:

    Status = HashPeImageByType (AuthData, AuthDataSize);
    if (EFI_ERROR (Status)) {
      continue;
    }

If the original behavior is a bug, then it should be fixed in a separate
patch.

--*--

For v4, I propose the following patch series structure:

- 1/8: Plug the leak described in comment (5).

- 2/8: Code flow update as described in comment (14) -- although it
       would be better to drop that change altogether!

- 3/8: (*CertBuffer) nullity check as described in comment (17)

- 4/8: If the assignment described in comment (26) is necessary, then
       dedicate a patch to it.

- 5/8: IsCertHashFoundInDatabase() updates, including call site updates
       in IsForbiddenByDbx() and IsAllowedByDb().

       Please address comments (1), (2), (3), (4), (6), (18) and (21).

- 6/8: Remaining updates to IsForbiddenByDbx().

       Please address comments (15), (16), and (19).

- 7/8: Remaining updates to IsAllowedByDb().

       Please address comment (20).

- 8/8: IsSignatureFoundInDatabase() updates, including call site updates
       in DxeImageVerificationHandler().

       Please address comments (7), (8), (9), (11), (12), (13), (22),
       (23), (24) and (25).

Thanks
Laszlo

&gt;|        }
&gt;|      }
&gt;|    }

**Comment by tianocore-issues** [comment]
## Comment 9444
Date: 2019-09-03 08:06:23 +0000
From: @lersek

(In reply to Laszlo Ersek from comment #18)

&gt; &gt;| @@ -1449,11 +1466,20 @@ IsAllowedByDb (
&gt; &gt;|                             mImageDigestSize
&gt; &gt;|                             );
&gt; &gt;|            if (VerifyStatus) {
&gt; &gt;| +            VerifyStatus = FALSE;
&gt; &gt;| +
&gt; &gt;|              //
&gt; &gt;|              // Here We still need to check if this RootCert&apos;s Hash is revoked
&gt; &gt;|              //
&gt; &gt;| +            DbxDataSize = 0;
&gt; &gt;|              Status   = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DbxDataSize, NULL);
&gt; &gt;| -            if (Status == EFI_BUFFER_TOO_SMALL) {
&gt; &gt;| +            if (Status != EFI_BUFFER_TOO_SMALL) {
&gt; &gt;| +              if (Status == EFI_NOT_FOUND) {
&gt; &gt;| +                //
&gt; &gt;| +                // Apparently not in dbx if the database doesn&apos;t exist.
&gt; &gt;| +                //
&gt; &gt;| +                VerifyStatus = TRUE;
&gt; &gt;| +              }
&gt; &gt;|                goto Done;
&gt; &gt;|              }
&gt; 
&gt; (20) I&apos;d again propose the pattern from (12) and (15).

Meaning two things:
- &quot;apparently&quot; vs. &quot;evidently&quot;,
- and the ASSERT().

The &quot;goto&quot; is justified in this case.

**Comment by tianocore-issues** [comment]
## Comment 9603
Date: 2019-09-11 04:46:42 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #18 &amp; #19)
Thank you very much for the very thorough review. I agree most of the comments. I&apos;ll give separate response later. v4 is being prepared and will bed uploaded soon.

**Comment by tianocore-issues** [comment]
## Comment 9626
Date: 2019-09-11 16:30:34 +0000
From: @lersek

Thanks, Jian! Please take your time. I will probably be slow to review v4; my plate has been way too full lately. Thanks.

**Comment by tianocore-issues** [comment]
## Comment 9634
Date: 2019-09-12 00:00:04 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #19)
About the ASSERT(), there&apos;re more similar cases in the same file. Do you think we need to add it to them all? But, on a second thought, I don&apos;t see any benefit from it. It seems no chances that GetVariable() will return EFI_SUCCESS in the case of variable buffer with zero size.

**Comment by tianocore-issues** [comment]
## Comment 9638
Date: 2019-09-12 03:11:28 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #19)
&gt;|              DbxData = (UINT8 *) AllocateZeroPool (DbxDataSize);
&gt;| @@ -1466,7 +1492,8 @@ IsAllowedByDb (
&gt;|                goto Done;
&gt;|              }
&gt;|
&gt;| -            if (IsCertHashFoundInDatabase (RootCert, RootCertSize, (EFI_SIGNATURE_LIST *)DbxData, DbxDataSize, &amp;RevocationTime)) {
&gt;| +            Status = IsCertHashFoundInDatabase (RootCert, RootCertSize, (EFI_SIGNATURE_LIST *)DbxData, DbxDataSize, &amp;RevocationTime, &amp;IsFound);
&gt;| +            if (!EFI_ERROR (Status) &amp;&amp; IsFound) {
&gt;|                //
&gt;|                // Check the timestamp signature and signing time to determine if the RootCert can be trusted.
&gt;|                //
&gt;| @@ -1474,6 +1501,8 @@ IsAllowedByDb (
&gt;|                if (!VerifyStatus) {
&gt;|                  DEBUG ((DEBUG_INFO, &quot;DxeImageVerificationLib: Image is signed and signature is accepted by DB, but its root cert failed the timestamp check.\n&quot;));
&gt;|                }
&gt;| +            } else if (!EFI_ERROR (Status) &amp;&amp; !IsFound) {
&gt;| +              VerifyStatus = TRUE;
&gt;|              }
&gt;|
&gt;|              goto Done;

Similar to (5), there&apos;s another memory leak here in multiple cert case. DbxData should be freed if it&apos;s not NULL before allocation. But I think we could move GetVariable related code to be out of while-loop, because the code is just retrieving the same variable and there&apos;s no need to read it again and again. But I&apos;m not sure if there&apos;s chance that it&apos;ll be updated elsewhere while being retrieved here. Possible?

**Comment by tianocore-issues** [comment]
## Comment 9640
Date: 2019-09-12 03:13:57 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Wang, Jian J from comment #23)
Sorry I quoted wrong code in last comment. See below

&gt;| @@ -1449,11 +1466,20 @@ IsAllowedByDb (
&gt;|                             mImageDigestSize
&gt;|                             );
&gt;|            if (VerifyStatus) {
&gt;| +            VerifyStatus = FALSE;
&gt;| +
&gt;|              //
&gt;|              // Here We still need to check if this RootCert&apos;s Hash is revoked
&gt;|              //
&gt;| +            DbxDataSize = 0;
&gt;|              Status   = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DbxDataSize, NULL);
&gt;| -            if (Status == EFI_BUFFER_TOO_SMALL) {
&gt;| +            if (Status != EFI_BUFFER_TOO_SMALL) {
&gt;| +              if (Status == EFI_NOT_FOUND) {
&gt;| +                //
&gt;| +                // Apparently not in dbx if the database doesn&apos;t exist.
&gt;| +                //
&gt;| +                VerifyStatus = TRUE;
&gt;| +              }
&gt;|                goto Done;
&gt;|              }

**Comment by tianocore-issues** [comment]
## Comment 9642
Date: 2019-09-12 12:22:18 +0000
From: @lersek

(In reply to Wang, Jian J from comment #22)
&gt; (In reply to Laszlo Ersek from comment #19)
&gt; About the ASSERT(), there&apos;re more similar cases in the same file. Do you
&gt; think we need to add it to them all? But, on a second thought, I don&apos;t see
&gt; any benefit from it. It seems no chances that GetVariable() will return
&gt; EFI_SUCCESS in the case of variable buffer with zero size.

My thinking was that the code deserved a comment -- it would really help my understanding at least if a comment stated explicitly that GetVariable() was bound to fail at that point (due to size=0).

And then, an ASSERT() is always better than a comment. :) It has zero cost in RELEASE builds, and provides extra safety in DEBUG/NOOPT builds.

Up to you.

**Comment by tianocore-issues** [comment]
## Comment 9643
Date: 2019-09-12 12:52:29 +0000
From: @lersek

(In reply to Wang, Jian J from comment #23)

&gt; Similar to (5), there&apos;s another memory leak here in multiple cert
&gt; case. DbxData should be freed if it&apos;s not NULL before allocation. But
&gt; I think we could move GetVariable related code to be out of
&gt; while-loop, because the code is just retrieving the same variable and
&gt; there&apos;s no need to read it again and again. But I&apos;m not sure if
&gt; there&apos;s chance that it&apos;ll be updated elsewhere while being retrieved
&gt; here. Possible?

Believe it or not :), I spent at least half an hour on that question!

So, it&apos;s a bit tricky. Note that, once we enter the &quot;VerifyStatus&quot;
block, we *always* jump to the &quot;Done&quot; label after one iteration. Please
see the &quot;goto&quot; statement that is at the very end of the block that is
conditional on &quot;VerifyStatus&quot;.

That is, there is no way to leak &quot;DbxData&quot;, because:

- &quot;VerifyStatus&quot; is FALSE, and we don&apos;t allocate &quot;DbxData&quot; at all, or
- &quot;VerifyStatus&quot; is TRUE, we allocate &quot;DbxData&quot;, but then we *alway*
  jump out of the loop (to the &quot;Done&quot; label).

My thoughts then turned to a more difficult question: was it indeed the
right thing to immediately terminate the loop (with success or failure)
once &quot;VerifyStatus&quot; is set to TRUE?

I mean sure, if the DBX check does *not* fire, then we&apos;re all good.

But in case the DBX check fires -- meaning that the current certificate,
selected from DB, accepts the executable, but the certificate is
blacklisted --, should we perhaps continue scanning DB for *another*
certificate that also accepts the binary, but is not rejected by DBX?

The UEFI spec answers this question, apparently. In &quot;32.5.3.3
Authorization Process&quot; in UEFI-2.8, we have

&gt; 3. The security database db must either contain an entry with a hash
&gt;    value of the image (with a supported hash type), or it must contain
&gt;    an entry with a certificate against which an entry in the image&apos;s
&gt;    certificate table can be verified. In either case verification must
&gt;    not succeed if the security database dbx contains any record with:
&gt;
&gt;    [...]
&gt;
&gt;    B. Any entry with SignatureListType of EFI_CERT_X509_SHA256,
&gt;       EFI_CERT_X509_SHA384, or EFI_CERT_X509_SHA512, with any
&gt;       SignatureData which reflects the To-Be-Signed hash included in
&gt;       any certificate in the signing chain of the signature being
&gt;       verified.
&gt;
&gt;    [...]

&quot;Any certificate in the signing chain of the signature being verified&quot;
seems to include &quot;current RootCert&quot;. Therefore the &quot;goto Done&quot; at the
end of the block appears justified to me.

And so DbxData cannot be leaked.

(

It&apos;s also interesting to look at IsAllowedByDb() historically.

- The &quot;goto Done&quot; statemenet originates from commit 0c18794ea428 (&quot;Add
  security package to repository.&quot;, 2011-09-02). At that time, the
  function was still called VerifyCertPkcsSignedData(), and it only
  searched for a match in DB -- it didn&apos;t consider the DBX blacklist.

- Then in commit 27c93c06c0a8 (&quot;Correct the DBX and Certificate Hash
  Checking.&quot;, 2015-02-04), the blacklist check was added. It extended
  the IsAllowedByDb() function with the current logic -- importantly, it
  *did not continue* the loop if the DBX check flipped VerifyStatus back
  to FALSE! Even if the DBX check set VerifyStatus to FALSE, the commit
  would allow the loop to terminate, like before, just because we found
  an entry in DB that accepted the binary.

  And based on the UEFI spec, I think it was intentional.

)

**Comment by tianocore-issues** [comment]
## Comment 9644
Date: 2019-09-12 13:00:48 +0000
From: @lersek

(I mean I had spent half an hour, dedicated to DbxData, when originally writing comment 18. That comment was going to be incredibly long anyway, and I determined DbxData was not possible to leak, plus the loop termination on DBX match looked intentional as well, so I decided to simply not comment on any of that. :) )

**Comment by tianocore-issues** [comment]
## Comment 9663
Date: 2019-09-14 04:40:31 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

For ASSERT(), I got your point. It&apos;s ok for me if it&apos;s for comments purpose. So let&apos;s keep it.

For DbxData, you&apos;re right. I neglected the &quot;goto Done&quot; part.

Thanks again for the great comments.

**Comment by tianocore-issues** [comment]
## Comment 9676
Date: 2019-09-16 17:59:44 +0000
From: @corthon

While we&apos;re discussing leaks. We&apos;ve identified one in IsAllowedByDb(), in the primary &quot;for&quot; loop, DbxData is allocated and not freed before the end of the loop. The final allocation is freed at the end of the function, but any intermediate allocations are leaked.

**Comment by tianocore-issues** [comment]
## Comment 9677
Date: 2019-09-16 18:05:53 +0000
From: @corthon

*OR* I could add a comment before reading all the updates where Laszlo diligently described why DbxData can&apos;t be leaked.

I would encourage a comment about why the leak is impossible. I would also love to see this code refactored in a way that this behavior is more apparent.

**Comment by tianocore-issues** [comment]
## Comment 9679
Date: 2019-09-16 21:55:22 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Bret Barkelew from comment #30)
&gt; *OR* I could add a comment before reading all the updates where Laszlo
&gt; diligently described why DbxData can&apos;t be leaked.
&gt; 
&gt; I would encourage a comment about why the leak is impossible. I would also
&gt; love to see this code refactored in a way that this behavior is more
&gt; apparent.

I think we can move those code (get dbx variable) to be outside of the while-loop. If no objection, I&apos;ll put change in the patch (1) in v4 series.

**Comment by tianocore-issues** [comment]
## Comment 9699
Date: 2019-09-17 05:53:21 +0000
From: @lersek

The code movement sounds good to me, thank you. I also agree that the final &quot;goto Done&quot; (i.e. &quot;VerifyStatus==TRUE&quot; being final) deserves a comment.

**Comment by tianocore-issues** [comment]
## Comment 9743
Date: 2019-09-20 02:11:33 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Created attachment 401
v4

v4 is uploaded. Please note that the logic in IsAllowedByDb() is also refined a little bit besides moving the GetVariable out of while-loop and adding comments.

Attachment: [BZ-401-v4.zip](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-401-v4.zip)

**Comment by tianocore-issues** [comment]
## Comment 9744
Date: 2019-09-20 02:14:21 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Wang, Jian J from comment #33)
&gt; Created attachment 401 [details]
&gt; v4
&gt; 
&gt; v4 is uploaded. Please note that the logic in IsAllowedByDb() is also
&gt; refined a little bit besides moving the GetVariable out of while-loop and
&gt; adding comments.

Patch 2/8 and 4/8 mentioned are not included because I agree to drop those changes.

**Comment by tianocore-issues** [comment]
## Comment 9745
Date: 2019-09-20 05:32:50 +0000
From: @lersek

Thanks, Jian. I&apos;ve queued this update for review. It will take a while; my backlog is crazy. :(

**Comment by tianocore-issues** [comment]
## Comment 9807
Date: 2019-09-25 05:26:35 +0000
From: @lersek

Jian,

I&apos;m very sorry, I must have missed your specific remark in comment 31 about *squashing* the movement of the dbx variable fetch into patch#1. In comment 32, I meant to agree to the code movement in general, and to a code comment on &quot;goto Done&quot;, in v4 -- but I certainly didn&apos;t mean to agree to doing that in patch#1, specifically.

Patch#1 -- fixing the leak in IsCertHashFoundInDatabase() -- is completely unrelated to the code movement. We *absolutely* need to separate the code movement to a stand-alone patch. In IsCertHashFoundInDatabase(), we fix a leak. In IsAllowedByDb(), we move around code to improve the readability of the function, but we do not change the behavior.

Again, the fault is with me, because you did announce your intent; I apologize for missing your intent. (It clearly happened because I&apos;ve been continuously overloaded for weeks or months now.) I should have spoken up earlier.

I have some more comments on v4 1/6, specifically in relation to EFI_BUFFER_TOO_SMALL.

- Inverting the EFI_BUFFER_TOO_SMALL condition at the top of the function (related to &quot;db&quot;) is helpful. It does not change behavior (there is no bug), but it simplifies the code flow, and it allows us to un-nest the rest of the function by one indentation level. Obviously, this patch should stand on its own.

- The EFI_BUFFER_TOO_SMALL change, related to &quot;dbx&quot;, is justified; it is a bugfix. It belongs in its own separate patch, together with the DbxDataSize=0 assignment.

So we need to split v4 1/6 into *at least* four patches:
- fixing the leak in IsCertHashFoundInDatabase(), like I requested in comment 18,

- the db:EFI_BUFFER_TOO_SMALL refactoring / un-indentation for IsAllowedByDb(),

- the dbx fetching fix (DbxDataSize=0 assignment, and EFI_BUFFER_TOO_SMALL check); *in-place*

- moving the dbx fetching out of the loop,

- (possibly commenting on &quot;goto Done&quot; for improving readability).

In its current form, v4 1/6 is practically un-reviewable to me.

--*--

I request that you please start minimalizing patches (*any* patches you write) as much as you can, as a *primary* goal. I shouldn&apos;t have to request this every single time I review a patch; I&apos;m asking for this as a *general* development method, a general attitude towards code changes. All patches should be as minimal as we can possibly make them.

Edk2 contributors really have to optimize the patches for the review process, and not for the programming (authoring) process.

- There is just one author, there are multiple reviewers.

- Writing code is easier than reading code. The author translates ideas to code, the reviewers need to translate code to ideas. The second is a lot more difficult.

If given a choice between twenty small patches and six messy patches, I&apos;ll take the former *any day*. I can stop and resume reviewing a  series at a patch boundary at any time. So if you break the patch set into twenty small patches, I&apos;ll have an easy time reviewing any given patch in the series, and I&apos;ll have twenty possible review suspension points, where I can go away and come back. If you give me 6 messy patches, I&apos;ll have to spend half a day *deciphering* each, and it destroys my mood and productivity.

Simplicity is the *most important* factor in software engineering. Every patch author must *actively* look for patch minimization opportunities, as a standing imperative, and vigorously exploit those opportunities.

**Comment by tianocore-issues** [comment]
## Comment 9808
Date: 2019-09-25 05:43:33 +0000
From: @lersek

For patch v4 2/6 (SecurityPkg/DxeImageVerificationLib: Add nullity check for CertBuffer):

I have now looked into this in more detail than in bullet (17).

The patch is valid, but the commit message is wrong.

The condition that we intend to catch here is whether the &quot;CertNumber&quot; field (of type UINT8) at the start of the output &quot;CertStack&quot; is zero.

Namely, after ensuring that BufferLength is positive, and CertBuffer is not NULL, the expression (*CertBuffer) indeed gives us the embedded &quot;CertNumber&quot; field. Therefore the commit message should say, &quot;catch if CertStack.CertNumber is zero&quot;.

However: this change is unnecessary! (Note that in bullet (17), I wrote that I couldn&apos;t tell whether this change was necessary.) Now I claim that this change is unnecessary.

Namely, just below we have:

  CertNumber = (UINT8) (*CertBuffer);
  CertPtr    = CertBuffer + 1;
  for (Index = 0; Index &lt; CertNumber; Index++) {

If CertNumber is 0, then the loop body is never entered, and we end up at the &quot;Done&quot; label.

So I think this change (patch v4 2/6) can be dropped altogether.

(Side comment: this is a good example why *small* patches are important. Now that in v4 this change is isolated to its own patch, I could investigate it separately from everything else, and arrive at a conclusion.)

**Comment by tianocore-issues** [comment]
## Comment 9809
Date: 2019-09-25 05:53:37 +0000
From: @lersek

Let me revise that: patch v4 2/6 has no effect on the control flow, with (*CertBuffer) == 0, but it changes the return value of the function.

So I was partly wrong; the patch is not superfluous.

However, the commit message is still wrong. The right message would be something like:

&quot;&quot;&quot;
SecurityPkg/DxeImageVerificationLib: reject CertStack.CertNumber==0 per DBX

In case the signers&apos; certificate stack, retrieved from the PE/COFF image&apos;s Authenticode blob, has zero elements (= there are zero signer certificates), then we should consider the image forbidden by DBX, not accepted by DBX.
&quot;&quot;&quot;

With this commit message update, for v4 2/6:

Reviewed-by: Laszlo Ersek &lt;lersek@redhat.com&gt;

**Comment by tianocore-issues** [comment]
## Comment 9810
Date: 2019-09-25 05:56:46 +0000
From: @lersek

(Meta comment: if my insistence on &quot;atomic&quot; patches is perceived as counter-productive, I&apos;m OK to withdraw from the review on this BZ. I certainly don&apos;t intend to slow down or block progress, if other reviewers on the infosec group would like to provide R-b&apos;s. Thanks.)

**Comment by tianocore-issues** [comment]
## Comment 9813
Date: 2019-09-25 08:03:29 +0000
From: @lersek

Created attachment 411
proposed restructuring of v4 patches 3/6 and 4/6

For patches v4 3/6 and v4 4/6:

I think we could further improve the scope of these patches.

* In patch v4 3/6 (&quot;SecurityPkg/DxeImageVerificationLib: Differentiate
  error and search result in IsCertHashFoundInDatabase&quot;), there is no
  need to flip the default value of IsForbidden to TRUE just yet, in
  IsForbiddenByDbx(). This patch is about updating the
  IsCertHashFoundInDatabase() function, and its call sites. By flipping
  the default value of IsForbidden to TRUE, in IsForbiddenByDbx(), we
  temporarily regress the EFI_NOT_FOUND interpretation for dbx. And then
  the next patch has to fix that up.

  So I would move the initial &quot;IsForbidden = TRUE&quot;, and the final
  &quot;IsForbidden = FALSE&quot;, to patch v4 4/6, for function
  IsForbiddenByDbx(). The commit message on patch v4 4/6 would have to
  be rewritten accordingly. (It would basically be the 2nd paragraph of
  v3 4/6:

      All intermediate results inside this function will be checked and
      returned immediately upon any failure or error, like
      out-of-resource, hash calculation error or certificate retrieval
      failure.

  )

* Furthermore, the leak fix from patch v4 4/6 should stand alone. (The
  code in question as also part of v3 -- please see it just under my
  bullet (16) in comment 18.)

Basically I&apos;m suggesting to replace the following two patches:

- v4 3/6 SecurityPkg/DxeImageVerificationLib: Differentiate error and
         search result in IsCertHashFoundInDatabase

- v4 4/6 SecurityPkg/DxeImageVerificationLib: Add logic to handle
         non-existence of dbx in IsForbiddenByDbx

With the following three patches:

- SecurityPkg/DxeImageVerificationLib: Differentiate error and search
  result in IsCertHashFoundInDatabase

  [updated]

- SecurityPkg/DxeImageVerificationLib: tighten default result of
  IsForbiddenByDbx()

  [new patch]

- SecurityPkg/DxeImageVerificationLib: plug Data leak in
  IsForbiddenByDbx()

  [new patch]

The end result is exactly the same, but the patch scopes would improve,
in my opinion.

Please see my proposal attached.

If you disagree, I&apos;m willing to give R-b for v4 3/6 and v4 4/6.

Thanks.

Attachment: [BZ-411-proposed_restructuring_of_v4_patches_3_and_4.zip](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-411-proposed_restructuring_of_v4_patches_3_and_4.zip)

**Comment by tianocore-issues** [comment]
## Comment 9830
Date: 2019-09-26 06:08:00 +0000
From: @lersek

I&apos;ve tried to look at patch v4 5/6
(&quot;SecurityPkg/DxeImageVerificationLib: Add logic to handle non-existence
of dbx in IsAllowedByDb&quot;).

It is difficult to make comments on this patch. The main problem is that
patch v4 1/6 regresses IsAllowedByDb(), for the case when DBX does not
exist.

Consider:

(a) After v4 patch#1 is applied, if DBX does not exist, IsAllowedByDb()
returns FALSE.

&gt;|   VerifyStatus      = FALSE;
&gt;|
&gt;|   [...]
&gt;|
&gt;|   DbxDataSize = 0
&gt;|   Status      = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DbxDataSize, NULL);
&gt;|   if (Status != EFI_BUFFER_TOO_SMALL) {
&gt;|     goto Done;
&gt;|   }

(b) However, at current edk2 master (c70fef962e80) -- that is, without
applying any patches from this BZ -- , we have the following code in
IsAllowedByDb():

&gt;|           if (VerifyStatus) {
&gt;|             //
&gt;|             // Here We still need to check if this RootCert&apos;s Hash is revoked
&gt;|             //
&gt;|             Status   = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DbxDataSize, NULL);
&gt;|             if (Status == EFI_BUFFER_TOO_SMALL) {
&gt;|               goto Done;
&gt;|             }
&gt;|             DbxData = (UINT8 *) AllocateZeroPool (DbxDataSize);
&gt;|             if (DbxData == NULL) {
&gt;|               goto Done;
&gt;|             }
&gt;|
&gt;|             Status = gRT-&gt;GetVariable (EFI_IMAGE_SECURITY_DATABASE1, &amp;gEfiImageSecurityDatabaseGuid, NULL, &amp;DbxDataSize, (VOID *) DbxData);
&gt;|             if (EFI_ERROR (Status)) {
&gt;|               goto Done;
&gt;|             }

When entering this block, the value in &quot;DbxDataSize&quot; is indeterminate
(due to a separate bug, discussed earlier). However, for the &quot;DBX does
not exist&quot; case, it actually doesn&apos;t matter.

VariableServiceGetVariable()
[MdeModulePkg/Universal/Variable/RuntimeDxe/Variable.c] calls
FindVariable() *first* -- which can return EFI_NOT_FOUND --, and does
the following only *second*, in accordance with the UEFI spec:

&gt;|   if (*DataSize &gt;= VarDataSize) {
&gt;|     if (Data == NULL) {
&gt;|       Status = EFI_INVALID_PARAMETER;
&gt;|       goto Done;
&gt;|     }

Which means that, if DBX does not exist, the first GetVariable()
invocation quoted under (b) will return EFI_NOT_FOUND, regardless of the
indeterminate value in DbxDataSize.

The EFI_BUFFER_TOO_SMALL check is wrong right after (discussed
elsewhere). Thus, we reach AllocateZeroPool().

Assume that AllocateZeroPool() fails (due to the indeterminate
DbxDataSize being impossible to allocate). Then we jump to Done, with
VerifyStatus=TRUE. This happens to be a *correct* result for the
function!

Assume that AllocateZeroPool() succeeds. Then we reach the last
GetVariable() call, for DBX. This will again fail with EFI_NOT_FOUND. We
jump to Done, with VerifyStatus=TRUE. This is again a *correct* result
for the function!

Therefore, despite all its flaws, IsAllowedByDb() at current master
(c70fef962e80) handles the total absence of DBX *correctly*.

And patch v4 1/6 breaks that, and then patch v4 5/6 restores it
separately. That&apos;s not a correct approach; if possible, we should not
break the &quot;DBX absent&quot; case for IsAllowedByDb() even temporarily.

I made some observations on v4 1/6 in comment#36 already. As I
mentioned, the DBX fetching changes in that patch must be broken out to
several smaller patches. And now I&apos;m adding: while at it, make sure that
the &quot;DBX absent&quot; return value for IsAllowedByDb() is never regressed,
relative to master.

With that in mind, patch v4 5/6 -- which depends on v4 1/6 -- will
either fall away completely, or it will have to be reworked.

**Comment by tianocore-issues** [comment]
## Comment 9831
Date: 2019-09-26 06:56:25 +0000
From: @lersek

Re: patch v4 6/6 (&quot;SecurityPkg/DxeImageVerificationLib: Differentiate error and search result in IsSignatureFoundInDatabase&quot;):

Reviewed-by: Laszlo Ersek &lt;lersek@redhat.com&gt;

**Comment by tianocore-issues** [comment]
## Comment 9969
Date: 2019-10-09 04:26:28 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #36)
Totally agree. I&apos;ll split the patches at suggested. Thanks for the very good advice.

**Comment by tianocore-issues** [comment]
## Comment 9970
Date: 2019-10-09 04:30:40 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #37)
&gt; For patch v4 2/6 (SecurityPkg/DxeImageVerificationLib: Add nullity check for
&gt; CertBuffer):
&gt; 
Good catch. Let&apos;s drop patch#2.

**Comment by tianocore-issues** [comment]
## Comment 9971
Date: 2019-10-09 04:34:57 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #38)
&gt; Let me revise that: patch v4 2/6 has no effect on the control flow, with
&gt; (*CertBuffer) == 0, but it changes the return value of the function.
&gt; 

Right. I almost forgot why :(, and wrote a wrong commit message.

**Comment by tianocore-issues** [comment]
## Comment 9972
Date: 2019-10-09 04:38:59 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #39)
&gt; (Meta comment: if my insistence on &quot;atomic&quot; patches is perceived as
&gt; counter-productive, I&apos;m OK to withdraw from the review on this BZ. I
&gt; certainly don&apos;t intend to slow down or block progress, if other reviewers on
&gt; the infosec group would like to provide R-b&apos;s. Thanks.)

I don&apos;t think it&apos;ll block anything. You did a very good review works. I really appreciate that. Forgive my slow response. I&apos;m really occupied by many other things.

**Comment by tianocore-issues** [comment]
## Comment 9973
Date: 2019-10-09 04:40:42 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #40)
&gt; Created attachment 411 [details]
&gt; proposed restructuring of v4 patches 3/6 and 4/6
&gt; 
&gt; For patches v4 3/6 and v4 4/6:
&gt; 

Totally agree.

**Comment by tianocore-issues** [comment]
## Comment 9974
Date: 2019-10-09 05:07:21 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #41)
&gt; I&apos;ve tried to look at patch v4 5/6
&gt; (&quot;SecurityPkg/DxeImageVerificationLib: Add logic to handle non-existence
&gt; of dbx in IsAllowedByDb&quot;).
&gt; 

It&apos;s brain burning comment:). I read it several times. But I still don&apos;t follow.  Forgive my numbness. I agree your analysis of DBX non-existence situation. But if DBX exists and DbxDataSize happens to be a small enough integer, the first gRT-&gt;GetVariable() do return EFI_BUFFER_TOO_SMALL. The original code did a &apos;==&apos; comparing and jump to &apos;Done&apos; if true. Isn&apos;t it still bug here?

            if (Status == EFI_BUFFER_TOO_SMALL) {
              goto Done;
            }

Do you mean we just need to fix above one but no need to init DbxDataSize?

**Comment by tianocore-issues** [comment]
## Comment 9978
Date: 2019-10-09 17:22:35 +0000
From: @lersek

Jian,

in comment 41, I focused only on the &quot;DBX absent&quot; case.

I did not consider any &quot;DBX present, and [insert other circumstances]&quot; situation.

My point was that, if DBX is absent, then patch#1 regresses the return value of IsAllowedByDb(), and then patch#5 restores the behavior.

Generally, when we work towards fixing a bug (or a set of bugs), we avoid introducing new ones, even temporarily (that is, even mid-series).

Such transitory bugs do not play nice with bisection. A person could be bisecting an independent issue, and the temporary breakage of IsAllowedByDb(), when DBX is absent, could prevent them from testing (interpreting) the symptom they were actually looking for, in the bisection. Put differently, such transitory bugs may leave a person with untestable commit ranges during a bisection.

So my request was that we please write patches that never break things, only refactor code (without changing behavior), or improve behavior.

Now I&apos;m fully aware that this is sometimes very difficult to do with small surgical steps, if at all possible :(

**Comment by tianocore-issues** [comment]
## Comment 9991
Date: 2019-10-10 01:54:06 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #49)
Understood. I think refactor will always cause bisection in this special case. Maybe we should not do it. But, from another point of view, I think there&apos;s also false-negative issue here. In original code, any error in fetching dbx will not alter the result of VerifyStatus (which is TRUE). Only timestamp check failure will change it to FALSE.

It means dbx timestamp check can be bypassed for other errors (like out-of-resource) even if it&apos;s expired. Similar to the issue in IsSignatureFoundInDatabase reported in tracker, any error in checking dbx should be taken as checking failure (not-allowed). I think we should also fix it here.

If you agree it&apos;s a false-negative issue in IsAllowedByDb(), I think it&apos;s OK to move the code fetching dbx to be outside the while-loop, because it&apos;s ok to return FALSE upon any error in fetching dbx. 

Otherwise, refactor this part of code will always cause bisection, because at that point (in refactored code) db has not been checked and we can only return FALSE upon errors (returning TRUE is a worse choice), which is different from original code logic.

**Comment by tianocore-issues** [comment]
## Comment 9992
Date: 2019-10-10 04:13:24 +0000
From: @lersek

In the end, I agree with you. It&apos;s more important to fix the &quot;incorrectly accepted&quot; situation than to keep the &quot;correctly accepted&quot; situation intact at every stage.

It looks like we cannot avoid a temporary &quot;incorrectly rejected&quot; behavior, midway into the series (when DBX is absent). But given the circumstances, I think it&apos;s tolerable, temporarily. Being too strict is not optimal, but if we have to choose either/or, it&apos;s better than being too lax. Please go ahead.

**Comment by tianocore-issues** [comment]
## Comment 9993
Date: 2019-10-10 04:31:46 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Created attachment 418
v5

Thanks, Laszlo. I uploaded v5 patches.

Attachment: [BZ-418-v5.zip](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-418-v5.zip)

**Comment by tianocore-issues** [comment]
## Comment 10023
Date: 2019-10-15 03:52:39 +0000
From: @lersek

Jian, please don&apos;t wait for my feedback on v5 -- please ask the owning package maintainers to review, and/or other reviewers from the infosec team. As I wrote elsewhere, I&apos;m ramping down my triage &amp; review activities in infosec. Thank you, and my apologies.

**Comment by tianocore-issues** [comment]
## Comment 10819
Date: 2020-01-09 11:55:04 +0000
From: @lersek

*** Bug 2450 has been marked as a duplicate of this bug. ***

**Comment by tianocore-issues** [comment]
## Comment 11072
Date: 2020-02-05 12:20:49 +0000
From: @vincent-j-zimmer

2/5/2020 infosec meeting - per starting time of embargo in june 19 and fact that we have patches + cve, move to un-embargo

**Comment by tianocore-issues** [comment]
## Comment 11110
Date: 2020-02-05 15:09:05 +0000
From: @lersek

Jian, please post the patches to the mailing list immediately, for maintainer review. This is a 0day that has been opened now, without patches being ready for immediate merging into upstream master.

**Comment by tianocore-issues** [comment]
## Comment 11111
Date: 2020-02-05 15:21:17 +0000
From: @lersek

In other words, we do *not* have patches, for upstream edk2. We would have patches for upstream edk2, and therefore for open source downstreams of edk2, if those patches were now ready for immediate posting to edk2-devel *and merging* into edk2. But prior review in this BZ had never reached that point. When I burned out and stepped away from the infosec team (comment 53), noone else joined this review, not even SecurityPkg maintainers. So, we do not have approved (ready-to-go) upstream patches.

**Comment by tianocore-issues** [comment]
## Comment 11123
Date: 2020-02-06 09:24:27 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Patches posted https://edk2.groups.io/g/devel/message/53866
Branch: https://github.com/jwang36/edk2/tree/fix-bz1608-bypass-blacklist-check-via-signature

**Comment by tianocore-issues** [comment]
## Comment 11262
Date: 2020-02-14 09:31:13 +0000
From: @lersek

* Jian posted:
[edk2-devel] [PATCH v2 00/10] Fix false negative issue in DxeImageVerificationHandler

* Mailing list archive links:
http://mid.mail-archive.com/20200214072745.1570-1-jian.j.wang@intel.com
https://edk2.groups.io/g/devel/message/54416

* Topic branch:
https://github.com/jwang36/edk2/tree/fix-bz1608-bypass-blacklist-check-via-signature-v2

**Comment by tianocore-issues** [comment]
## Comment 11302
Date: 2020-02-19 09:21:58 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Pushed fbb9607223...c230c002ac

**Comment by tianocore-issues** [comment]
## Comment 17359
Date: 2021-09-22 15:47:02 +0000
From: Kim Olsun &lt;<kimolsun2020>&gt;

http://www.compilatori.com/category/tech/	http://www.wearelondonmade.com/category/tech/	http://www.jopspeech.com/category/tech/	http://joerg.li/category/tech/	http://connstr.net/category/tech/	http://embermanchester.uk/category/tech/	http://www.slipstone.co.uk/category/tech/	http://www.logoarts.co.uk/category/tech/	http://www.acpirateradio.co.uk/category/tech/	https://waytowhatsnext.com/category/property/	https://www.webb-dev.co.uk/category/property/	http://www.iu-bloomington.com/category/property/	http://www-look-4.com/category/tech/
